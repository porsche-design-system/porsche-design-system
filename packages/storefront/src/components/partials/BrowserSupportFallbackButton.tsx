'use client';

import { isDevEnvironment } from '@/utils/isDev';
import { PButton } from '@porsche-design-system/components-react/ssr';
import { CDN_BASE_PATH, FALLBACKS_MANIFEST } from '@porsche-design-system/fallbacks';

export const showBrowserSupportFallback = () => {
  const script = document.createElement('script');
  const src = `https://cdn.ui.porsche.com${CDN_BASE_PATH}/${FALLBACKS_MANIFEST.browserSupport}`;
  script.src = isDevEnvironment
    ? src.replace('https://cdn.ui.porsche.com/porsche-design-system', 'http://localhost:3001')
    : src;
  document.body.appendChild(script);
};

export const BrowserSupportFallbackButton = () => {
  return <PButton onClick={() => showBrowserSupportFallback()}>Force display of browser support fallback</PButton>;
};
