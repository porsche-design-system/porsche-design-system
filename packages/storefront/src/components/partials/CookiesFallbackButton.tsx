'use client';

import { isDevEnvironment } from '@/utils/isDev';
import { PButton } from '@porsche-design-system/components-react/ssr';
import { CDN_BASE_PATH, FALLBACKS_MANIFEST } from '@porsche-design-system/fallbacks';

export const showCookiesFallback = () => {
  const script = document.createElement('script');
  const src = `https://cdn.ui.porsche.com${CDN_BASE_PATH}/${FALLBACKS_MANIFEST.cookies}`;
  script.src = isDevEnvironment
    ? src.replace('https://cdn.ui.porsche.com/porsche-design-system', 'http://localhost:3001')
    : src;
  document.body.appendChild(script);
};

export const CookiesFallbackButton = () => {
  return <PButton onClick={() => showCookiesFallback()}>Force display of cookies fallback</PButton>;
};
