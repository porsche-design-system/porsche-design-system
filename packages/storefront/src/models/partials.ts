export type Partials =
  | 'getBrowserSupportFallbackScript'
  | 'getComponentChunkLinks'
  | 'getCookiesFallbackScript'
  | 'getDSRPonyfill'
  | 'getFontFaceStyles'
  | 'getFontFaceStylesheet'
  | 'getFontLinks'
  | 'getIconLinks'
  | 'getInitialStyles'
  | 'getLoaderScript'
  | 'getMetaTagsAndIconLinks';
export type PartialLocation = 'head' | 'body';

export type PartialCall = {
  params: PartialParam[];
  comment?: string;
};

export type PartialParam = {
  key: string;
  value: string | string[];
};
