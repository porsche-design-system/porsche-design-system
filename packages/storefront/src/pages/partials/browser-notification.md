# Notifications

<TableOfContents></TableOfContents>

## Browser Notification

### Install

The Browser Notification partials are provided in the `@porsche-design-system/browser-notification`. 

It's necessary to have access to the Porsche Design System private npm registry to be able to install the `@porsche-design-system/browser-notification` npm package. If you don't have an account yet, please first [read more about getting started as developer](start-coding/introduction).

```
// install with npm:
npm install @porsche-design-system/browser-notification

// install with yarn:
yarn add @porsche-design-system/browser-notification
```

### Usage

We provide two notifications variants to show the user the corresponding information in regard to its used browser.

Include the **Browser Notification** into your project by importing and calling the provided `includeBanner()` or `includeOverlay()` function within your `index.html` just before the closing `</body>` tag (requires a bundler like Webpack, Rollup or a small Node JS script). This adds a `<script>` tag with pre-defined browser- and feature detection pointing to the corresponding browser notification JS snippet hosted on a CDN.

#### Banner notification

The **Banner** variant is meant to inform the user if they access the application with a browser which doesn't get full support by the application. The banner is placed above the page and can be closed by the user.

Integration through `includeBanner()` partial.

| Browser Detection           | Initialized |
| --------------------------- | ----------- |
| **MS Edge(HTML) <= 18**     | ✓           |
| **IE <= 11**                | ✓           |
| **Microsoft Edge Chromium** | ✗           |

#### Overlay notification

The **Overlay** variant is meant to inform the user when they access the application if their browser doesn't support **custom elements**
or browser technologies like [**IntersectionObserver**](https://caniuse.com/?search=intersectionobserver) and [**MutationObserver**](https://caniuse.com/?search=mutationobserver) which are required to render the Porsche Design System components.
The overlay is placed above the page, covers it completely and can't be closed or removed by the user.

Integration through `includeOverlay()` partial.

| Feature Detection                                                                               | Initialized |
| ------------------------------------------------------------------------------------------------| ----------- |
| **Custom elements != true**                                                                     | ✓           |
| **IntersectionObserver != true**                                                                | ✓           |
| **MutationObserver != true**                                                                    | ✓           |
| **Custom elements && IntersectionObserver && MutationObserver**                                 | ✗           |

#### React / Vue JS

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>Some title</title>
  </head>
  <body>
    <div id="app"></div>

    <!-- inline version of the banner or overlay init script -->
    <%= require('@porsche-design-system/browser-notification').includeBanner() %>
    <%= require('@porsche-design-system/browser-notification').includeOverlay() %>
  </body>
</html>
```

#### Angular / Vanilla JS

```
// index.html
<body>
  <!--PLACEHOLDER-->
</body>

// package.json
"scripts": {
  "banner": "partial=$(node -e 'console.log(require(\"@porsche-design-system/browser-notification\").includeBanner().replace((\\\\[bd\\/]|&)/g, \"\\\\$1\"))') && regex='<!--PLACEHOLDER-->|<script>.*browser-notification.*<\\/script>' && sed -i'' -E -e \"s@$regex@$partial@\" index.html",
  "overlay": "partial=$(node -e 'console.log(require(\"@porsche-design-system/browser-notification\").includeOverlay().replace((\\\\[bd\\/]|&)/g, \"\\\\$1\"))') && regex='<!--PLACEHOLDER-->|<script>.*browser-notification.*<\\/script>' && sed -i'' -E -e \"s@$regex@$partial@\" index.html"
}
```

### Translations

Automatic translations for the following languages are provided:  
`'de' | 'ru' | 'fr' | 'en' | 'it' | 'pt' | 'es' | 'ja' | 'ko' | 'zh' | 'nl' | 'pl' | 'cs' | 'da' | 'et' | 'fi' | 'lt' | 'lv' | 'no' | 'sl' | 'sv' | 'tr' | 'uk'`

The **Browser Notification** is looking once as soon as the script initializes for the obligatory `lang` attribute defined in the `html` tag. Support is given for the following formats, e.g.:

- `lang="en"`
- `lang="en_US"`
- `lang="en-US"`

If none of these languages can be found, it will fall back to `en`.

### Troubleshooting

There always might be a case where something goes wrong. Here are some possible answers:

1. **Q:** Why does the translation not get recognized automatically?  
   **A:** Mostly this is a result of false order of script loading and setting translation key by the application. It's required that the `lang` attribute in the `html` tag is defined with the correct value before the **Browser Notification** script gets initialized.  
   **A:** The translation key is not part of the provided keys (see "Translations")  
   **A:** The translation key has not the correct format (see "Translations")
2. **Q:** Why are there no implementation guidelines for my JS framework?  
   **A:** Implementing a third party script can be done in many ways regarding the setup of your application. So there isn't a solely true way to integrate it in a specific framework. Just one rule of thumb: **It should be initialized as late as possible.**
