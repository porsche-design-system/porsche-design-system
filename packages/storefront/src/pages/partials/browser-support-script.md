# Browser Support Script

**Function name:** `getBrowserSupportScript()`

The Porsche Design System components rely on dedicated JavaScript features like `IntersectionObserver`,
`MutationObserver`, `customElements` and others. If any of the requirements are not met, a blocking overlay is displayed
which recommends the user to update the browser.

Therefore, we provide a ready to use partial in all `@porsche-design-system/components-{js|angular|react}` packages
which needs to be injected before the closing `</body>` of your `index.html`.

## Supported options

| Option   | Description                                                                                                               | Type    | Default |
| -------- | ------------------------------------------------------------------------------------------------------------------------- | ------- | ------- | -------- |
| `cdn`    | Decides from which CDN the resources are loaded.                                                                          | `'auto' | 'cn'`   | `'auto'` |
| `format` | Defines the output format of the partial. By default, it returns a html string, with `jsx` it returns valid jsx elements. | `'html' | 'jsx'`  | `'html'` |

## Examples

Project integration differs based on the project setup.  
The following showcases the most common ways.

<PartialDocs name="getBrowserSupportScript" :params="params" location="body"></PartialDocs>

<script lang="ts">
import Vue from 'vue';
import Component from 'vue-class-component';

@Component
export default class Code extends Vue {
  public params = [
    {
      value: ""
    },
    {
      value: "{ cdn: 'cn' }",
      comment: 'force using China CDN',
    },
  ];
}
</script>
