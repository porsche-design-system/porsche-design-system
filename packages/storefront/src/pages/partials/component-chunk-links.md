# Component Chunk Links

**Function name:** `getComponentChunkLinks()`

Porsche Design System components load dynamically from a CDN as soon as they are used for the first time.  
This results in a waterfall like loading behaviour where your application bootstraps first, then loads the Porsche
Design System Core and when any component rendered the corresponding component chunk gets loaded afterwards.  
This can be optimized by fetching used chunks in parallel while the application is bootstrapping.

Therefore, we provide a ready to use partial in all `@porsche-design-system/components-{js|angular|react|vue}` packages
which needs to be injected into the `<head>` of your `index.html`.

An in-depth optimization guide can be found at
[Vanilla Js Optimization](must-know/initialization/vanilla-js#optimization).

## Supported options

| Option       | Description                                                                                                                                                                                                        | Type                           | Default |
| ------------ | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | ------------------------------ | ------- | -------- | -------- |
| `components` | All components listed in the array are loaded from the CDN. By default, our core is always preloaded when using this partial.                                                                                      | `({{this.componentChunks}})[]` | `[]`    |
| `cdn`        | Decides from which CDN the resources are loaded.                                                                                                                                                                   | `'auto'                        | 'cn'`   | `'auto'` |
| `format`     | Defines the output format of the partial. By default, it returns a html string, with `jsx` it returns valid jsx elements. When using the <code class="no-before">js</code> option a javascript object is returned. | `'html'                        | 'jsx'   | 'js'`    | `'html'` |

When using `jsx` in the `format` option, it is necessary to have `react/jsx-runtime` as a dependency in the project
included.

## Examples

Project integration differs based on the project setup.  
The following showcases the most common ways.

**Note:** Make sure to preload only component chunks which are really needed initially!

<PartialDocs name="getComponentChunkLinks" :params="params" location="head"></PartialDocs>

<script lang="ts">
import Vue from 'vue';
import Component from 'vue-class-component';
import { COMPONENT_CHUNK_NAMES } from '../../../../components-js/projects/components-wrapper';

@Component
export default class Code extends Vue {
  public componentChunks = COMPONENT_CHUNK_NAMES.map(x => `'${x}'`).join(' | ');
  public params = [
    {
      value: "{ components: ['button', 'marque'] }"
    },
    {
      value: "{ cdn: 'cn' }",
      comment: 'force using China CDN'
    },
  ];
}
</script>
