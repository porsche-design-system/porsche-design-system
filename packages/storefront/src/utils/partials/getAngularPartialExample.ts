import type { PartialCall, PartialLocation, Partials } from '@/models/partials';
import { formatPartialParams } from '@/utils/partials/formatPartialParams';

export const getAngularPartialExample = (name: Partials, location: PartialLocation, partialCalls: PartialCall[]) => {
  const partialImportPath = '@porsche-design-system/components-angular/partials';
  const glue = '\n  ';

  const angularPartials = partialCalls
    .map(({ comment, params }, i) => {
      return [
        comment && `// ${comment}`,
        `${i === 0 ? 'let ' : ''}partialContent = ${name}(${formatPartialParams(params)});`,
      ]
        .filter(Boolean)
        .join(glue);
    })
    .join('\n\n  ');
  return `<!-- prerequisite -->
<!-- docs: https://github.com/just-jeb/angular-builders/tree/master/packages/custom-webpack#index-transform -->
yarn add --dev @angular-builders/custom-webpack

<!-- angular.json -->
...
"architect": {
  "build": {
-   "builder": "@angular-devkit/build-angular:browser",
+   "builder": "@angular-builders/custom-webpack:browser",
    "options": {
      "outputPath": "dist/components-angular",
+     "indexTransform": "./scripts/injectPartials.ts",
    }
  }
  "serve": {
-   "builder": "@angular-devkit/build-angular:dev-server",
+   "builder": "@angular-builders/custom-webpack:dev-server",

<!-- ./scripts/injectPartials.ts -->
import type { TargetOptions } from '@angular-builders/custom-webpack';
import { ${name} } from '${partialImportPath}';

export default (targetOptions: TargetOptions, indexHtml: string): string => {
  ${angularPartials}

  return indexHtml.replace(/<\\/${location}>/, \`\${partialContent}$&\`);
};`;
};
