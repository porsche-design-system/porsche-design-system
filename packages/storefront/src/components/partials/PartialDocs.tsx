import { CodeBlock } from '@/components/playground/CodeBlock';
import type { PartialCall, PartialLocation, Partials } from '@/models/partials';
import { getAngularPartialExample } from '@/utils/partials/getAngularPartialExample';
import { getNextPartialExample } from '@/utils/partials/getNextPartialExample';
import { getReactPartialExample } from '@/utils/partials/getReactPartialExample';
import { getVanillaJsPartialExample } from '@/utils/partials/getVanillaJsPartialExample';
import { getVuePartialExample } from '@/utils/partials/getVuePartialExample';
import type React from 'react';
import SyntaxHighlighter from 'react-syntax-highlighter';
import { H3, P } from '../../../mdx-components';

type PartialDocsProps = {
  name: Partials;
  location: PartialLocation;
  partialCalls: PartialCall[];
};

export const PartialDocs = ({ name, location, partialCalls }: PartialDocsProps) => {
  // Apply some basic formatting to make the output easier readable
  const formatPartial = (partial: unknown): string => {
    if (typeof partial === 'string') {
      return partial.replace(/(>)/g, '>\n').replace(/(<\/)/g, '\n</');
    }
    return String(partial);
  };

  return (
    <>
      <CodeBlock
        frameworkMarkup={{
          'vanilla-js': getVanillaJsPartialExample(name, location, partialCalls),
          angular: getAngularPartialExample(name, location, partialCalls),
          react: getReactPartialExample(name, location, partialCalls),
          vue: getVuePartialExample(name, location, partialCalls),
          next: getNextPartialExample(name),
        }}
      />
      <H3>Output</H3>
      <P>The result of this partial looks like this:</P>
      {/* @ts-expect-error: Suppress type incompatibility */}
      <SyntaxHighlighter
        className="markup select-none"
        language="html"
        useInlineStyles={false}
        codeTagProps={{ tabIndex: 0 }}
      >
        {partialCalls
          .map(({ comment, params }) => {
            const paramObj = Object.fromEntries(params.map(({ key, value }) => [key, value]));
            const partialResult = require('@porsche-design-system/components-react/partials')[name](paramObj);
            return `${comment ? `// ${comment}\n` : ''}${formatPartial(partialResult)}`;
          })
          .join('\n')}
      </SyntaxHighlighter>
    </>
  );
};
