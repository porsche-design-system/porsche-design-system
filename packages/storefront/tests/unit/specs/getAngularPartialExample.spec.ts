import { getAngularPartialExample } from '@/utils/partials/getAngularPartialExample';
import { expect } from 'vitest';

it('should return correct partial markup for getVanillaJsPartialExample', () => {
  const partialCalls = [
    {
      params: [],
    },
    {
      comment: 'Alternative: With custom prefix to match your prefixed components',
      params: [
        {
          key: 'prefix',
          value: 'custom-prefix',
        },
      ],
    },
    {
      comment: 'Alternative: With multiple prefixes to match prefixed components coming from micro frontends',
      params: [
        {
          key: 'prefix',
          value: ['', 'custom-prefix', 'another-prefix'],
        },
      ],
    },
  ];
  const partialExample = getAngularPartialExample('getInitialStyles', 'head', partialCalls);
  expect(partialExample).toMatchSnapshot();
});
