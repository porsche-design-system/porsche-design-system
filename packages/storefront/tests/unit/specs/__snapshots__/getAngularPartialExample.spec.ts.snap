// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should return correct partial markup for getVanillaJsPartialExample 1`] = `
"<!-- prerequisite -->
<!-- docs: https://github.com/just-jeb/angular-builders/tree/master/packages/custom-webpack#index-transform -->
yarn add --dev @angular-builders/custom-webpack

<!-- angular.json -->
...
"architect": {
  "build": {
-   "builder": "@angular-devkit/build-angular:browser",
+   "builder": "@angular-builders/custom-webpack:browser",
    "options": {
      "outputPath": "dist/components-angular",
+     "indexTransform": "./scripts/injectPartials.ts",
    }
  }
  "serve": {
-   "builder": "@angular-devkit/build-angular:dev-server",
+   "builder": "@angular-builders/custom-webpack:dev-server",

<!-- ./scripts/injectPartials.ts -->
import type { TargetOptions } from '@angular-builders/custom-webpack';
import { getInitialStyles } from '@porsche-design-system/components-angular/partials';

export default (targetOptions: TargetOptions, indexHtml: string): string => {
  let partialContent = getInitialStyles();

  // Alternative: With custom prefix to match your prefixed components
  partialContent = getInitialStyles({ prefix: 'custom-prefix' });

  // Alternative: With multiple prefixes to match prefixed components coming from micro frontends
  partialContent = getInitialStyles({ prefix: ['', 'custom-prefix', 'another-prefix'] });

  return indexHtml.replace(/<\\/head>/, \`\${partialContent}$&\`);
};"
`;
